File: src/server/lib/atomic-write.ts
import { randomBytes } from "node:crypto";
import { renameSync, unlinkSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";

export function writeFileAtomic(
  filePath: string,
  content: string,
  encoding: BufferEncoding = "utf-8",
): void {
  const dir = dirname(filePath);
  const tmpPath = join(dir, `.tmp-${randomBytes(6).toString("hex")}`);
  try {
    writeFileSync(tmpPath, content, encoding);
    renameSync(tmpPath, filePath);
  } catch (err) {
    try {
      unlinkSync(tmpPath);
    } catch {}
    throw err;
  }
}

File: server/lib/atomicwrite.go
package lib

import (
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"os"
	"path/filepath"
)

// WriteFileAtomic writes content to filePath atomically (write-to-temp, rename).
// content is []byte (callers should convert strings with []byte(s)).
// The file is written with permissions 0o644 regardless of the existing file's permissions.
// Ensures no partial writes are visible to concurrent readers.
func WriteFileAtomic(filePath string, content []byte) error {
	dir := filepath.Dir(filePath)
	b := make([]byte, 6)
	if _, err := rand.Read(b); err != nil {
		return fmt.Errorf("atomicwrite rand: %w", err)
	}
	tmpPath := filepath.Join(dir, ".tmp-"+hex.EncodeToString(b))

	if err := os.WriteFile(tmpPath, content, 0o644); err != nil {
		return fmt.Errorf("atomicwrite write tmp: %w", err)
	}

	if err := os.Rename(tmpPath, filePath); err != nil {
		_ = os.Remove(tmpPath)
		return fmt.Errorf("atomicwrite rename: %w", err)
	}

	return nil
}
